### PR TITLE
Fixed CCX coach dashboard access issues

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
### Background
fixes https://github.com/mitocw/edx-platform/issues/245, https://github.com/mitocw/edx-platform/issues/246

### What is done in this PR
**Studio Updates:** None.
**LMS Updates:** 
 - 404 if user access coach dashboard with invalid CCX id in url.
 - 404 if enable_ccx flag is set to false in studio course settings.

@pdpinch @aliceriot  @gsidebo 

##### In case of ccx is disable for course: 
- If you have valid ccx then disable enable_ccx settings from advance settings in studio then use url and try to access ccx coach dashboard. you will get 404 exception.

##### In case of ccx id in url is invalid:
- Modify ccx coach dashboard URL in browser change ccx id (number) to invalid number let say 700 and hit, you will get 404 error.

#### IF enable_ccx flag is set to false:
screen shot 2016-06-26 at 11 05 26 pm

#### IF coach dashboard access with invalid ccx id in url:
screen shot 2016-06-26 at 11 05 59 pm

### CODE REVIEW

**Comment by pdpinch** [comment]
Can we use the standard 404 page instead of this custom message?

**Comment by pdpinch** [comment]
I think it would be safer and more readable to check for `enable_ccx` before looking for a `ccx_id`

In other words, move https://github.com/mitocw/edx-platform/pull/247/files#diff-faebd8eb1d591f3ad3147fe858e4e849R100-103 before https://github.com/mitocw/edx-platform/pull/247/files#diff-faebd8eb1d591f3ad3147fe858e4e849R87

**Comment by amir-qayyum-khan** [comment]
@pdpinch 

- At that point we dont have actual course object thats why we can check enable_ccx
- Because we have to see if course key is ccx course key then we need to extract master course key from it otherwise the key is master course id.

**Comment by pdpinch** [comment]
@giocalitri can you take a look at this?

**Comment by giocalitri** [comment]
sure

File: lms/djangoapps/ccx/views.py [hunk 1]
```diff
@@ -84,14 +85,19 @@ def wrapper(request, course_id):
         ccx = None
         if isinstance(course_key, CCXLocator):
             ccx_id = course_key.ccx
-            ccx = CustomCourseForEdX.objects.get(pk=ccx_id)
-            course_key = ccx.course_id
+            try:
+                ccx = CustomCourseForEdX.objects.get(pk=ccx_id)
+                course_key = ccx.course_id
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: please move this line after the `try/except`

File: lms/djangoapps/ccx/views.py [hunk 2]
```diff
@@ -84,14 +85,19 @@ def wrapper(request, course_id):
...
+            except CustomCourseForEdX.DoesNotExist:
+                raise Http404
 
         course = get_course_by_id(course_key, depth=None)
         is_staff = has_access(request.user, 'staff', course)
         is_instructor = has_access(request.user, 'instructor', course)
 
-        if is_staff or is_instructor:
+        if not course.enable_ccx:
+            raise Http404
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: when does this case happen? when a CCX has been created and then the CCX option disabled in advanced settings?

**Comment by giocalitri** [comment]
two very small comments

File: lms/djangoapps/ccx/views.py (hunk 2 above)
Comment (amir-qayyum-khan) [inline lms/djangoapps/ccx/views.py]: yes this happens when

- CCX option disabled in advanced settings
- You have existing ccx 
- After disable ccx in advance setting you access you valid ccx via url

Comment (amir-qayyum-khan) [inline lms/djangoapps/ccx/views.py]: Also updated the description of PR and added a block how to test.

**Comment by amir-qayyum-khan** [comment]
Done @giocalitri

File: lms/djangoapps/ccx/tests/test_views.py [hunk 3]
```diff
@@ -187,6 +187,15 @@ class TestCoachDashboard(CcxTestCase, LoginEnrollmentTestCase):
     Tests for Custom Courses views.
     """
 
+    @classmethod
+    def setUpClass(cls):
+        super(TestCoachDashboard, cls).setUpClass()
+        cls.course_disable_ccx = CourseFactory.create(enable_ccx=False)
+        cls.course_with_ccx_coonect_set = CourseFactory.create(
```
Comment (giocalitri) [inline lms/djangoapps/ccx/tests/test_views.py]: I think this is a typo: `coonect` in `cls.course_with_ccx_coonect_set`

Comment (amir-qayyum-khan) [inline lms/djangoapps/ccx/tests/test_views.py]: fixed @giocalitri

**Comment by giocalitri** [comment]
👍

**Comment by giocalitri** [comment]
please squash before opening the PR against EDX

**Comment by amir-qayyum-khan** [comment]
Moved PR on edX [upstream](https://github.com/edx/edx-platform/pull/12889)

**Comment by pdpinch** [comment]
Ok. I was about to confirm that I did the manual testing, and everything looked good. 

@giocalitri, would you mind adding a thumb to the upstream PR?

**Comment by giocalitri** [comment]
there are some issues there with quality, but I will give it as soon as everything is green

**Pull request closed without merging**